rust/ip-packet/src/lib.rs

```diff
@@ -0,0 +1,377 @@
+        } else {
+            return;
+        };
+
+        self.as_tcp()
+            .expect("Developer error: we can only get a checksum if the packet is tcp")
+            .set_checksum(checksum);
+    }
+
+    pub fn into_immutable(self) -> IpPacket<'a> {
+        match self {
+            Self::Ipv4(p) => p.consume_to_immutable().into(),
+            Self::Ipv6(p) => p.consume_to_immutable().into(),
+        }
+    }
+
+    pub fn as_immutable(&self) -> IpPacket<'_> {
+        match self {
+            Self::Ipv4(p) => IpPacket::Ipv4(p.to_immutable()),
+            Self::Ipv6(p) => IpPacket::Ipv6(p.to_immutable()),
+        }
+    }
+
+    pub fn as_udp(&mut self) -> Option<MutableUdpPacket> {
+        self.to_immutable()
+            .is_udp()
+            .then(|| MutableUdpPacket::new(self.payload_mut()))
+            .flatten()
+    }
+
+    fn as_tcp(&mut self) -> Option<MutableTcpPacket> {
+        self.to_immutable()
+            .is_tcp()
+            .then(|| MutableTcpPacket::new(self.payload_mut()))
+            .flatten()
+    }
+
+    fn set_icmpv6_checksum(&mut self) {
+        let (src_addr, dst_addr) = match self {
+            MutableIpPacket::Ipv4(_) => return,
+            MutableIpPacket::Ipv6(p) => (p.get_source(), p.get_destination()),
+        };
+        if let Some(mut pkt) = self.as_icmpv6() {
+            let checksum = icmpv6::checksum(&pkt.to_immutable(), &src_addr, &dst_addr);
+            pkt.set_checksum(checksum);
+        }
+    }
+
+    fn as_icmpv6(&mut self) -> Option<MutableIcmpv6Packet> {
+        self.to_immutable()
+            .is_icmpv6()
+            .then(|| MutableIcmpv6Packet::new(self.payload_mut()))
+            .flatten()
+    }
+
+    pub fn as_immutable_udp(&self) -> Option<UdpPacket> {
+        self.to_immutable()
+            .is_udp()
+            .then(|| UdpPacket::new(self.payload()))
+            .flatten()
+    }
+
+    pub fn as_immutable_tcp(&self) -> Option<TcpPacket> {
+        self.to_immutable()
+            .is_tcp()
+            .then(|| TcpPacket::new(self.payload()))
+            .flatten()
+    }
+
+    pub fn swap_src_dst(&mut self) {
+        match self {
+            Self::Ipv4(p) => {
+                swap_src_dst!(p);
+            }
+            Self::Ipv6(p) => {
+                swap_src_dst!(p);
+            }
+        }
+    }
+
+    #[inline]
+    pub fn set_dst(&mut self, dst: IpAddr) {
+        match (self, dst) {
+            (Self::Ipv4(p), IpAddr::V4(d)) => p.set_destination(d),
+            (Self::Ipv6(p), IpAddr::V6(d)) => p.set_destination(d),
+            _ => {}
+        }
+    }
+
+    #[inline]
+    pub fn set_src(&mut self, src: IpAddr) {
+        match (self, src) {
+            (Self::Ipv4(p), IpAddr::V4(s)) => p.set_source(s),
+            (Self::Ipv6(p), IpAddr::V6(s)) => p.set_source(s),
+            _ => {}
+        }
+    }
+
+    pub fn set_len(&mut self, total_len: usize, payload_len: usize) {
```

Comment:
```suggestion
    /// Sets the length in the packet header without resizing the payload slice
    /// 
    /// Parameters:
    /// - `total_len` - Only used for IPv4, ignored for IPv6
    /// - `payload_len` - Ignored for IPv4, only used for IPv6
    ///
    /// If the length does not fit into a `u16`, then the length will be set incorrectly. See <https://doc.rust-lang.org/reference/expressions/operator-expr.html#numeric-cast>
    /// For performance reasons, we do not panic on failed cast to `u16`,
    /// and for ergonomic reasons, we must take `usize` parameters.
    pub fn set_len(&mut self, total_len: usize, payload_len: usize) {
```

This function looks _real_ weird to me so I guessed at the behavior and wrote a doc comment.

Reply:
Yeah I agree. I only copied it over but I think I'll just inline it instead which should make this more obvious.